File: packages/php/src/Seguranca/Ataques/AtaqueCiphertextEstatistico.php
<?php

namespace Application\Seguranca\Ataques;

use Application\Seguranca\AlvoCriptografico;
use Application\Seguranca\AtaqueInterface;
use Application\Seguranca\CriptografiaAlvo;
use Application\Seguranca\ResultadoAtaque;
use Application\Seguranca\SkipAtaqueException;

/**
 * Estatística só do CIPHERTEXT (não do keystream): distribuição de bytes
 * (qui-quadrado), teste de runs nos bits e autocorrelação lag-1. Um ciphertext
 * de cifra sólida deve parecer ruído mesmo com plaintext variado.
 */
class AtaqueCiphertextEstatistico implements AtaqueInterface
{
    public function __construct(
        private int $amostras = 200,
        private int $tamanhoTexto = 64,
        private float $zLimite = 4,
    ) {
    }

    public function nome(): string
    {
        return 'Estatística do ciphertext (chi²/runs/autocorrelação)';
    }

    public function executar(AlvoCriptografico $alvo): ResultadoAtaque
    {
        if (!$alvo instanceof CriptografiaAlvo) {
            throw new SkipAtaqueException('Precisa de base64url_encode/decode do alvo.');
        }

        $contagem = array_fill(0, 256, 0);
        $bytes = [];

        for ($s = 0; $s < $this->amostras; $s++) {
            $texto = random_bytes($this->tamanhoTexto);
            $token = $alvo->encrypt($texto);
            $campos = $alvo->decompor($alvo->base64urlDecode(substr($token, strlen($alvo->prefixo()))));

            $ct = $campos['ciphertext'];
            $lenCt = strlen($ct);
            for ($i = 0; $i < $lenCt; $i++) {
                $b = ord($ct[$i]);
                $contagem[$b]++;
                $bytes[] = $b;
            }
        }

        $total = count($bytes);
        $esperado = $total / 256;
        $chi2 = 0.0;
        for ($b = 0; $b < 256; $b++) {
            $d = $contagem[$b] - $esperado;
            $chi2 += ($d * $d) / $esperado;
        }

        // runs test nos bits do ciphertext
        $uns = 0;
        $bits = [];
        for ($i = 0; $i < $total; $i++) {
            $b = $bytes[$i];
            for ($k = 7; $k >= 0; $k--) {
                $bit = ($b >> $k) & 1;
                $bits[] = $bit;
                $uns += $bit;
            }
        }
        $n = count($bits);
        $pi = $uns / $n;
        $runs = 1;
        for ($i = 1; $i < $n; $i++) {
            if ($bits[$i] !== $bits[$i - 1]) {
                $runs++;
            }
        }
        $esperadoRuns = 2 * $n * $pi * (1 - $pi);
        $desvioRuns = 2 * sqrt(2 * $n) * $pi * (1 - $pi);
        $zRuns = $desvioRuns > 0 ? abs($runs - $esperadoRuns) / $desvioRuns : 0.0;

        // autocorrelação lag-1
        $media = array_sum($bytes) / $total;
        $num = 0.0;
        $den = 0.0;
        for ($i = 0; $i < $total; $i++) {
            $den += ($bytes[$i] - $media) ** 2;
        }
        for ($i = 0; $i < $total - 1; $i++) {
            $num += ($bytes[$i] - $media) * ($bytes[$i + 1] - $media);
        }
        $autocorr = $den > 0 ? $num / $den : 0.0;

        $problemas = [];
        if ($chi2 > 330) {
            $problemas[] = sprintf('qui-quadrado=%.1f (>330 suspeito)', $chi2);
        }
        if ($zRuns > $this->zLimite) {
            $problemas[] = sprintf('runs z=%.2f', $zRuns);
        }
        if (abs($autocorr) > 0.1) {
            $problemas[] = sprintf('autocorrelação=%.3f', $autocorr);
        }

        $vulneravel = !empty($problemas);

        return new ResultadoAtaque(
            $this->nome(),
            vulneravel: $vulneravel,
            severidade: $vulneravel ? 'media' : 'info',
            detalhes: $vulneravel
                ? implode('; ', $problemas)
                : sprintf(
                    'qui-quadrado=%.1f sobre %d bytes, runs z=%.2f, autocorrelação=%.3f - todos dentro do esperado',
                    $chi2,
                    $total,
                    $zRuns,
                    $autocorr,
                ),
            dados: ['chi2' => $chi2, 'runs_z' => $zRuns, 'autocorrelacao' => $autocorr],
        );
    }
}

File: packages/rust/src/seguranca/ataques/ataque_ciphertext_estatistico.rs
//! Estatística só do CIPHERTEXT (não do keystream): distribuição de bytes
//! (qui-quadrado), teste de runs nos bits e autocorrelação lag-1. Um ciphertext
//! de cifra sólida deve parecer ruído mesmo com plaintext variado.

use crate::seguranca::alvo_criptografico::AlvoCriptografico;
use crate::seguranca::ataque::{Ataque, ErroAtaque};
use crate::seguranca::resultado_ataque::{ResultadoAtaque, Severidade};
use crate::seguranca::util::random_bytes;

pub struct AtaqueCiphertextEstatistico {
    pub amostras: usize,
    pub tamanho_texto: usize,
    pub z_limite: f64,
}

impl AtaqueCiphertextEstatistico {
    pub fn new() -> Self {
        AtaqueCiphertextEstatistico {
            amostras: 200,
            tamanho_texto: 64,
            z_limite: 4.0,
        }
    }
}

impl Default for AtaqueCiphertextEstatistico {
    fn default() -> Self {
        Self::new()
    }
}

impl Ataque for AtaqueCiphertextEstatistico {
    fn nome(&self) -> String {
        "Estatística do ciphertext (chi²/runs/autocorrelação)".to_string()
    }

    fn executar(&self, alvo: &dyn AlvoCriptografico) -> Result<ResultadoAtaque, ErroAtaque> {
        let mut contagem = [0u64; 256];
        let mut bytes: Vec<u8> = Vec::new();

        for _ in 0..self.amostras {
            let texto = random_bytes(self.tamanho_texto);
            let token = alvo
                .encrypt(&texto)
                .map_err(|e| ErroAtaque::Falha(e.to_string()))?;
            let decodificado = alvo
                .base64url_decode(&token[alvo.prefixo().len()..])
                .map_err(|e| ErroAtaque::Falha(e.to_string()))?;
            let campos = alvo.decompor(&decodificado);
            for b in campos.ciphertext {
                contagem[b as usize] += 1;
                bytes.push(b);
            }
        }

        let total = bytes.len();
        let esperado = total as f64 / 256.0;
        let mut chi2 = 0.0f64;
        for &c in &contagem {
            let d = c as f64 - esperado;
            chi2 += d * d / esperado;
        }

        // runs test nos bits do ciphertext
        let mut uns = 0usize;
        let mut bits: Vec<u8> = Vec::with_capacity(total * 8);
        for &b in &bytes {
            for k in (0..8).rev() {
                let bit = (b >> k) & 1;
                bits.push(bit);
                uns += bit as usize;
            }
        }
        let n = bits.len();
        let pi = uns as f64 / n as f64;
        let mut runs = 1usize;
        for i in 1..n {
            if bits[i] != bits[i - 1] {
                runs += 1;
            }
        }
        let esperado_runs = 2.0 * n as f64 * pi * (1.0 - pi);
        let desvio_runs = 2.0 * (2.0 * n as f64).sqrt() * pi * (1.0 - pi);
        let z_runs = if desvio_runs > 0.0 {
            (runs as f64 - esperado_runs).abs() / desvio_runs
        } else {
            0.0
        };

        // autocorrelação lag-1
        let media = bytes.iter().map(|&b| b as f64).sum::<f64>() / total as f64;
        let mut num = 0.0f64;
        let mut den = 0.0f64;
        for &b in &bytes {
            den += (b as f64 - media).powi(2);
        }
        for i in 0..total.saturating_sub(1) {
            num += (bytes[i] as f64 - media) * (bytes[i + 1] as f64 - media);
        }
        let autocorr = if den > 0.0 { num / den } else { 0.0 };

        let mut problemas: Vec<String> = Vec::new();
        if chi2 > 330.0 {
            problemas.push(format!("qui-quadrado={:.1} (>330 suspeito)", chi2));
        }
        if z_runs > self.z_limite {
            problemas.push(format!("runs z={:.2}", z_runs));
        }
        if autocorr.abs() > 0.1 {
            problemas.push(format!("autocorrelação={:.3}", autocorr));
        }

        let vulneravel = !problemas.is_empty();

        Ok(ResultadoAtaque::new(
            &self.nome(),
            vulneravel,
            if vulneravel {
                Severidade::Media
            } else {
                Severidade::Info
            },
            &if vulneravel {
                problemas.join("; ")
            } else {
                format!(
                    "qui-quadrado={:.1} sobre {} bytes, runs z={:.2}, autocorrelação={:.3} - todos dentro do esperado",
                    chi2, total, z_runs, autocorr
                )
            },
        ))
    }
}
